src/core/atomic.d

```diff
@@ -112,39 +112,41 @@ TailShared!T atomicLoad(MemoryOrder ms = MemoryOrder.seq, T)( ref const shared T
  *  val    = The target variable.
  *  newval = The value to store.
  */
-void atomicStore(MemoryOrder ms = MemoryOrder.seq, T, V)( ref T val, V newval ) pure nothrow @nogc @trusted
-    if ( __traits( compiles, { val = newval; } ) && !is(T == shared S, S) && !is(V == shared U, U) )
+void atomicStore(MemoryOrder ms = MemoryOrder.seq, T, V)(ref T val, V newval) pure nothrow @nogc @trusted
+    if (!is(T == shared S, S) && !is(V == shared U, U))
 {
-    static if ( __traits(isFloating, T) )
+    // resolve implicit conversions
+    T arg = newval;
```

Comment:
Hmm, doesn't this end up `@trust`ing potentially arbitrary user code?

Reply:
That's an interesting point.
It's invalid for any type that interacts with atomic to have non-trivial copy anyway... so I can assert that in the constraint and then we can have confidence this is safe.
I actually thought about that while doing this (not for this exact reason), but it's stuck on https://issues.dlang.org/show_bug.cgi?id=19902, which is a really bad regression, and it's been sitting there for months!
I figured I'd add that restriction whenever this is solved.

Reply:
I put the assert.
I'm not sure if the same assert should be everywhere though; `atomicStore` is definitely a copying assignment, whereas exchange and cas are both actually swap operations, which are move operations, so they don't actually interact with copy semantics.

Reply:
@TurkeyMan: This line causes a regression for:
```
struct S
{
    int x;
    alias x this;
}
shared S s;
s = 0x1234_5678;
atomicStore(s, 0); // => Error: cannot implicitly convert expression `newval` of type `int` to `S`
assert(s.x == 0);
```

Note that the removed `__traits( compiles, { val = newval; } )` constraint doesn't play a role due to `alias this` magic, i.e., assignment is fine, but construction via implicit conversion fails. An explicit cast would work.

Reply:
interesting.
so you're saying the semantic difference is that i changed a copy to a construction?

is this real code? does this occur somewhere?

Reply:
It's only an additional LDC test (https://github.com/ldc-developers/druntime/blob/77066d7d399b0da1e94699da37436066589898ef/src/core/atomic.d#L1073-L1081), so I hope code like this isn't in the wild.

I'm not sure how to handle it best, i.e., whether such code should be disallowed or whether an explicit cast in `atomicStore` is justified.